ddl_checker/main.go

```diff
@@ -0,0 +1,209 @@
+		"You can switch modes using the `SETMOD` command.\n" +
+		"Auto mode: The program will automatically synchronize the dependent table structure from MySQL " +
+		"and delete the conflict table\n" +
+		"Prompt mode: The program will ask you before synchronizing the dependent table structure from MYSQL\n" +
+		"Offline mode: This program doesn't need to connect to MySQL, and doesn't perform anything other than executing the input SQL.\n\n" +
+		setmodUsage + "\n"
+
+	setmodUsage = "SETMOD usage: SETMOD <MODCODE>; MODCODE = [\"Auto\", \"Prompt\", \"Offline\"] (case insensitive).\n"
+
+	auto    = "auto"
+	prompt  = "prompt"
+	offline = "offline"
+)
+
+func main() {
+	fmt.Print(welcomeInfo)
+	initialise()
+	mainLoop()
+	destroy()
+}
+
+func initialise() {
+	flag.Parse()
+	var err error
+	reader = bufio.NewReader(os.Stdin)
+	executableChecker, err = checker.NewExecutableChecker()
+	if err != nil {
+		fmt.Printf("[DDLChecker] Init failed, can't create ExecutableChecker: %s\n", err.Error())
+		os.Exit(1)
+	}
+	executableChecker.Execute(tidbContext, "use test;")
+	dbInfo := &dbutil.DBConfig{
+		User:     *username,
+		Password: *password,
+		Host:     *host,
+		Port:     *port,
+		Schema:   *schema,
+	}
+	ddlSyncer, err = checker.NewDDLSyncer(dbInfo, executableChecker)
+	if err != nil {
+		fmt.Printf("[DDLChecker] Init failed, can't open mysql database: %s\n", err.Error())
+		os.Exit(1)
+	}
+}
+
+func destroy() {
+	executableChecker.Close()
+	ddlSyncer.Close()
+}
+
+func mainLoop() {
+	var input string
+	var err error
+	for isContinue := true; isContinue; isContinue = handler(input) {
+		fmt.Printf("[%s] > ", strings.ToTitle(mode))
+		input, err = reader.ReadString(';')
+		if err != nil {
+			fmt.Printf("[DDLChecker] Read stdin error: %s\n", err.Error())
+			os.Exit(1)
+		}
+	}
+}
+func handler(input string) bool {
+	lowerTrimInput := strings.ToLower(strings.TrimFunc(input, func(r rune) bool {
+		return unicode.IsSpace(r) || r == ';'
+	}))
+	// cmd exit
+	if lowerTrimInput == "exit" {
+		return false
+	}
+	// cmd setmod
+	if strings.HasPrefix(lowerTrimInput, "setmod") {
+		x := strings.TrimSpace(lowerTrimInput[6:])
+		switch x {
+		case auto, prompt, offline:
+			mode = x
+		default:
+			fmt.Print(setmodUsage)
+		}
+		return true
+	}
+	if mode != offline {
+		// auto and query mod
+		stmt, err := executableChecker.Parse(input)
+		if err != nil {
+			fmt.Printf("[DDLChecker] SQL parse error: %s\n", err.Error())
+			return true
+		}
+		neededTables, _ := checker.GetTablesNeededExist(stmt)
+		nonNeededTables, err := checker.GetTablesNeededNonExist(stmt)
+		// skip when stmt isn't a DDLNode
+		if err == nil && (mode == auto || (mode == prompt && promptAutoSync(neededTables, nonNeededTables))) {
+			err := syncTablesFromMysql(neededTables)
+			if err != nil {
+				return true
+			}
+			err = dropTables(nonNeededTables)
+			if err != nil {
+				return true
+			}
+		}
+	}
+	err := executableChecker.Execute(tidbContext, input)
```

Comment:
If I input a valid non-DDL SQL, the checker still tells me `SQL execution succeeded`. The tool is more like a `SQL checker` with a TiDB context. Maybe we can put some DDL check here and tell user whether the input SQL is a DDL or not, but this requires SQL parsing and some more work, which is not as simple as current code.

Reply:
i can warning users when they  input a valid non-DDL SQL and continue execute the sql.